middleware/definitions.go

```diff
@@ -0,0 +1,77 @@
+package middleware
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	_ "github.com/ory/jsonschema/v3/fileloader"
+	_ "github.com/ory/jsonschema/v3/httploader"
+	"github.com/ory/x/configx"
+	"github.com/ory/x/logrusx"
+
+	"github.com/ory/oathkeeper/driver"
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/proxy"
+	"github.com/ory/oathkeeper/rule"
+	"github.com/ory/oathkeeper/x"
+)
+
+type (
+	dependencies interface {
+		Logger() *logrusx.Logger
+		RuleMatcher() rule.Matcher
+		ProxyRequestHandler() proxy.RequestHandler
+	}
+
+	middleware struct{ dependencies }
+
+	options struct {
+		logger       *logrusx.Logger
+		configFile   string
+		registryAddr *driver.Registry
+	}
+
+	Option func(*options)
+)
+
+// WithLogger uses the specified logger in the middleware.
+func WithLogger(l *logrusx.Logger) Option {
+	return func(o *options) { o.logger = l }
+}
+
+// WithConfigFile sets the path to the config file to use for the middleware.
+func WithConfigFile(configFile string) Option {
+	return func(o *options) { o.configFile = configFile }
+}
+
+// New creates an Oathkeeper middleware from the options. By default, it tries
+// to read the configuration from the file "oathkeeper.yaml".
+func New(ctx context.Context, opts ...Option) (*middleware, error) {
+	o := options{
+		logger:     logrusx.New("Ory Oathkeeper Middleware", x.Version),
+		configFile: "oathkeeper.yaml",
+	}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	c, err := configuration.NewKoanfProvider(
+		ctx, nil, o.logger,
```

Comment:
Is the nil flag ok here? I'm not sure what impact it has

Reply:
The nil is the flags, so if used as a middleware Oathkeeper should not consume any flags, i.e., the `--config` flag.

This reminds me to check that it **also** does not consume environment variables. IMHO the only way to configure the middleware should be through the config file, so that no options clash between Oathkeeper and the main application.

Reply:
OK, now I also isolated the ENV. I think it's fine to completely disable env loading (as it is now), but we could also have env vars under a prefix `OATHKEEPER_`, if we really want env support in the middleware. WDYT?

Reply:
Ok, makes sense for now